### PR TITLE
Fix zero division error for empty dataloaders

### LINKED ISSUE #12830: Raise a error [ZeroDivisionError: float division by zero] at trainer/connectors/data_connector.py
### To Reproduce

Use a data loader with no data, i.e., `len(dataloader) == 0`.

### Expected behavior

Raise error `ZeroDivisionError: float division by zero`.

### Environment

- PyTorch Lightning Version (e.g., 1.5.0):  1.6.1
- PyTorch Version (e.g., 1.10): 1.10.2
- Python version (e.g., 3.9): 3.9.12
- OS (e.g., Linux): Windows
- CUDA/cuDNN version: CUDA 11.3
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source): conda
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### Additional context

Hey guys, PyTorchLightning has helped me simplify a lot of my model training work, and it's really great!

Today I ran into an error which `ZeroDivisionError: float division by zero`, after troubleshooting, I found was caused by this line of code:

https://github.com/PyTorchLightning/pytorch-lightning/blob/1233554e73303f4f3d30f9a9219e2aa8c3f6a467/pytorch_lightning/trainer/connectors/data_connector.py#L402

The path to my validation dataset was pointing to the wrong place, so `len(val_dataloader)` was 0.

So I'd like to add a check here to alert the user in a friendly way that its data loader has no data!

Thanks for all the efforts!

cc @borda @justusschock @awaelchli @ninginthecloud @rohitgr7

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #12830

Also if there are insufficient batches when using `limi_train_batches=float` (for eg, `total batches=9 and limit_train_batches=0.1`), it is used to silently disable the training. Ideally, it should raise an error just like it does with eval dataloaders.

Also, it is weird that when the input dataloaders have no batches, it just raises a warning and doesn't do anything but for the above case, it raises an exception. IMO we should start raising an exception in this case too?

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @justusschock @awaelchli @ninginthecloud @rohitgr7 @borda

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1860,8 +1866,19 @@ def reset_train_dataloader(self, model: Optional["pl.LightningModule"] = None) -
...
 
-        # store epoch of dataloader reset for reload_dataloaders_every_n_epochs
-        self._last_train_dl_reload_epoch = self.current_epoch
+        if (
+            self.num_training_batches == 0
+            and self.limit_train_batches > 0.0
+            and isinstance(self.limit_train_batches, float)
+            and orig_train_batches != float("inf")
+        ):
+            min_pct = 1.0 / orig_train_batches
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1875]: do you know what "pct" means?

File: tests/trainer/connectors/test_data_connector.py [hunk 2]
```diff
@@ -171,3 +172,17 @@ def val_dataloader(self):
     trainer._data_connector.attach_data(model)
     trainer.reset_val_dataloader(model)
     assert trainer.val_dataloaders[0].sampler.shuffle == shuffle
+
+
+def test_error_raised_with_insufficient_float_limit_train_dataloader():
+    batch_size = 16
+    dl = DataLoader(RandomDataset(32, batch_size * 9), batch_size=batch_size)
+    trainer = Trainer(limit_train_batches=0.1, limit_val_batches=0.1)
```
Comment (carmocca) [inline tests/trainer/connectors/test_data_connector.py:180]: Why set `limit_val_batches`?

**Review by carmocca — APPROVED** [review]
> Also, it is weird that when the input dataloaders have no batches, it just raises a warning and doesn't do anything but for the above case, it raises an exception. IMO we should start raising an exception in this case too?

I agree

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:1873]: percentage, I guess

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -1416,58 +1416,6 @@ def test_predict_return_predictions_cpu(return_predictions, precision, tmpdir):
         assert preds[0].dtype == (torch.float64 if precision == 64 else torch.float32)
 
 
-@pytest.mark.parametrize(
-    ["limit_train_batches", "global_step", "num_training_batches", "current_epoch", "should_train"],
-    [(0.2, 0, 0, 0, False), (0.5, 10, 2, 5, True)],
-)
-def test_disabled_training_for_insufficient_limit_train_batches(
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply tests/trainer/test_trainer.py:1423]: Check the description. Now we raise an error instead so this test is irrelevant.

File: tests/trainer/connectors/test_data_connector.py (hunk 2 above)
Comment (rohitgr7) [reply tests/trainer/connectors/test_data_connector.py:180]: ah! can be removed.

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 4]
```diff
@@ -381,36 +381,44 @@ def _reset_eval_dataloader(
...
+                    num_batches = min(orig_num_batches, limit_eval_batches)
+                elif isinstance(limit_eval_batches, float) and orig_num_batches != float("inf"):
+                    num_batches = int(orig_num_batches * limit_eval_batches)
                 elif limit_eval_batches != 1.0:
                     raise MisconfigurationException(
-                        f"When using an IterableDataset for `limit_{mode}_batches`,"
-                        f" `Trainer(limit_{mode.dataloader_prefix}_batches)` must be `1.0` or an int. An int k"
-                        f" specifies `num_{mode.dataloader_prefix}_batches` to use."
+                        f"When using an `IterableDataset`, `Trainer(limit_{mode.dataloader_prefix}_batches)` must be"
+                        f"`1.0` or an int. An int specifies `num_{mode.dataloader_prefix}_batches` to use."
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/data_connector.py:408]:
```suggestion
                        f" `1.0` or an int. An int specifies `num_{mode.dataloader_prefix}_batches` to use."
```

File: tests/trainer/test_trainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply tests/trainer/test_trainer.py:1423]: This is a breaking change then. Labelling it as such

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -1862,8 +1866,19 @@ def reset_train_dataloader(self, model: Optional["pl.LightningModule"] = None) -
...
+        if (
+            self.num_training_batches == 0
+            and self.limit_train_batches > 0.0
+            and isinstance(self.limit_train_batches, float)
+            and orig_train_batches != float("inf")
+        ):
+            min_percentage = 1.0 / orig_train_batches
+            raise MisconfigurationException(
+                f"You requested to check {self.limit_train_batches} of the `train_dataloader` but"
+                f" {self.limit_train_batches} * {orig_train_batches} < 1. Please increase the"
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:1878]:
```suggestion
                f" {self.limit_train_batches} * {orig_train_batches} = 1. Please increase the"
```
of can it be negative?

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:1878]: should be `< 1` no? the error is thrown for cases `[0, 1)`. Can't be negative.

Comment (otaj) [reply pytorch_lightning/trainer/trainer.py:1878]: I guess it can be between 0 and 1, since `limit_train_batches` can be a float

**Review by otaj — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Pull request merged**